### PR TITLE
Add auth refresh flow

### PR DESCRIPTION
**Pull request by anthonyhashemi**
## Changes in this PR

- Add auth refresh flow

## JIRA ticket

https://national-archives.atlassian.net/browse/AYR-420

### CODE REVIEW

File: app/main/routes.py [hunk 1]
```diff
@@ -82,6 +82,8 @@ def callback():
 
     session["access_token"] = access_token_response["access_token"]
     session["refresh_token"] = access_token_response["refresh_token"]
+    decoded_access_token = keycloak_openid.introspect(session["access_token"])
+    session["user_groups"] = decoded_access_token["groups"]
```
Comment (anthonyhashemi) [inline app/main/routes.py:86]: noticed that neither sign-in nor callback routes are covered by normal pytest flask tests, only by e2e playwright tests. The fact that all our permission based browse and search tests and the sign in tests work in e2e is a good coverage already but I may add a couple of simple tests for these routes in a followup PR which mock keycloack and assert on the session.

I will override the code cov check for this line because of this

File: app/main/authorize/access_token_sign_in_required.py [hunk 2]
```diff
@@ -40,10 +41,25 @@ def decorated_view(*args, **kwargs):
...
         try:
-            if not _validate_access_token():
+            access_token = session["access_token"]
+            refresh_token = session["refresh_token"]
+            try:
+                (
+                    access_token,
+                    refresh_token,
+                    tokens_are_refreshed,
+                ) = _validate_or_refresh_tokens(access_token, refresh_token)
```
Comment (anthonyhashemi) [inline app/main/authorize/access_token_sign_in_required.py:51]: _validate_or_refresh_tokens doesnt concern itself with sessions.

Did consider just testing all the edge cases in that function itself, making it a public function but after tidying up the test cases I think its just about okay how it is, if it got any bigger I would do that.

File: app/main/authorize/access_token_sign_in_required.py [hunk 3]
```diff
@@ -63,18 +79,30 @@ def decorated_view(*args, **kwargs):
     return decorated_view
 
 
-def _validate_access_token():
-    is_valid_access_token = False
+def _validate_or_refresh_tokens(access_token, refresh_token):
+    tokens_are_refreshed = False
+
+    if not (access_token and refresh_token):
```
Comment (anthonyhashemi) [inline app/main/authorize/access_token_sign_in_required.py:85]: I made the assumption that we should always have access_token and refresh_token set due to us setting them both in the callback and nobody can manipulate the session due to the flask secret.

This assumption keeps the test cases simpler

File: app/main/authorize/access_token_sign_in_required.py [hunk 4]
```diff
@@ -40,10 +41,25 @@ def decorated_view(*args, **kwargs):
...
+                (
+                    access_token,
+                    refresh_token,
+                    tokens_are_refreshed,
+                ) = _validate_or_refresh_tokens(access_token, refresh_token)
+            except InvalidAccessToken:
                 session.clear()
                 return redirect(url_for("main.sign_in"))
 
+            if tokens_are_refreshed:
```
Comment (anthonyhashemi) [inline app/main/authorize/access_token_sign_in_required.py:56]: no point to update the session if they werent refreshed, as theyd be the same values as before

File: app/main/routes.py (hunk 1 above)
Comment (pgandhizaizi) [reply app/main/routes.py:86]: I agree, we should have at least basic tests for code coverage for sign-in and callback routes

File: app/main/authorize/access_token_sign_in_required.py (hunk 3 above)
Comment (pgandhizaizi) [reply app/main/authorize/access_token_sign_in_required.py:85]: good and precise clarification, thank you

File: app/main/authorize/access_token_sign_in_required.py (hunk 4 above)
Comment (pgandhizaizi) [reply app/main/authorize/access_token_sign_in_required.py:56]: image

**Comment by anthonyhashemi** [comment]
Just realised that even with this solution, after the refresh token expires due to the client expiry time, once we are redirected back to the sign-in page, we are redirected back to the keycloak login page which seems to auto create a new token and allow us back in to the app.

Screenshot 2024-02-08 at 18 35 33

**Comment by anthonyhashemi** [comment]
> Just realised that even with this solution, after the refresh token expires due to the client expiry time, once we are redirected back to the sign-in page, we are redirected back to the keycloak login page which seems to auto create a new token and allow us back in to the app.
> 
> Screenshot 2024-02-08 at 18 35 33

Actually, after reading the keycloak documentation, I realise that it was how I tested this out that caused this issue. I had changed the client idle time but this is optional. Without it, the refresh token expires when the sso session expires, and I think that is fine for us, we don't need to log out a user from this client specifically.

**Review by pgandhizaizi — APPROVED** [review]
looks good, but we need to create a separate ticket to write test cases for sign in and callback route
